### PR TITLE
Implement distribute and zip (zipWith) for parsers

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

File: src/Streamly/Internal/Data/Parse.hs [hunk 1]
```diff
@@ -80,6 +88,23 @@ all predicate = Parse step initial (\x -> return $ Right (0,x))
             else Halt False
         else Halt False
 
+{-# INLINE sum #-}
+sum :: (Monad m, Num a) => Parse m a a
+sum = Parse (\s a -> return $ Keep 0 (s + a))   -- step
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parse.hs]: There is a `sum` in `Fold`. You can remove this as we have `fromFold`.

File: src/Streamly/Internal/Data/Parse.hs [hunk 2]
```diff
@@ -252,3 +277,101 @@ endBefore = undefined
...
 finishBy :: Parse m a x -> Fold m a y -> Parse m a (x, y)
 finishBy = undefined
+
+-------------------------------------------------------------------------------
+-- Zipping Parses
+-------------------------------------------------------------------------------
+--
+
+-- | State of the pair of parsers in a zip composition
+-- Note: strictness annotation is important for fusing the constructors
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parse.hs]: Is it the fusion of constructors or constructor elimination via specialization?

File: src/Streamly/Internal/Data/Parse.hs [hunk 3]
```diff
@@ -252,3 +277,101 @@ endBefore = undefined
...
+
+    {-# INLINE_LATE initial #-}
+    initial = do
+        sL <- initialL
+        sR <- initialR
+        return $ ParPair ([], Right sL, [], []) ([], Right sR, [], [])
+
+    {-# INLINE_LATE step #-}
+    step (ParPair (bufL, Right sL, inpL1, inpL2)
+                  (bufR, Right sR, inpR1, inpR2)) x = do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parse.hs]: What is `inpL1` and `inpL2` here?

Comment (adithyaov) [reply src/Streamly/Internal/Data/Parse.hs]: Won't `inpR2` and `inpL2` always be `[]`?

Comment (adithyaov) [reply src/Streamly/Internal/Data/Parse.hs]: I believe you are using these for efficiency? Similar to how d-list is implemented for efficient `snoc` and `reverse`?

File: src/Streamly/Internal/Data/Parse.hs [hunk 4]
```diff
@@ -252,3 +277,101 @@ endBefore = undefined
...
+    {-# INLINE extractBoth #-}
+    extractBoth sL sR = do
+        rL <- extractL sL
+        case rL of
+            Left err -> return $ Left err
+            Right (_, x) -> do
+                rR <- extractR sR
+                return $ case rR of
+                    Left err -> Left err
+                    Right (_, y) -> Right (0, zf x y)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parse.hs]: Is `Right (0 ...` correct?
Shouldn't it be more like `Right (min (length bufL1 + length bufL2) (length bufR1 + length bufR2)) ...`

**Comment by harendra-kumar** [comment]
I will push this again, there are several changes to the previous code.

File: src/Streamly/Internal/Data/Parse.hs (hunk 3 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/Parse.hs]: Cool, Cool!

**Pull request closed without merging**